A43-sdk-authorization.md

```diff
@@ -0,0 +1,364 @@
+			"type": "string",
+		},
+		"deny_rules": {
+			"description": "List of deny rules to match. If a request matches any of the deny rules, then it will be denied. If none of"
+			  "the deny rules matches or there are no deny rules, the allow rules will be evaluated.",
+			"type": "array",
+			"items": {
+				"$ref": "#/definitions/rule",
+			},
+		},
+		"allow_rules": {
+			"description": "List of allow rules to match. The allow rules will only be evaluated after the deny rules. If a request"
+			  "matches any of the allow rules, then it will allowed. If none of the allow rules matches, it will be denied.",
+			"type": "array",
+			"items": {
+				"$ref": "#/definitions/rule",
+			},
+		}
+	},
+	"required": ["name", "allow_rules"]
+}
+```
+
+gRPC SDK authorization policy has a list of allow rules and a list of deny rules. 
+The following sequence is followed to make an authorization decision -
+1. Check for a match in deny rules, if a match is found, the request will be denied. 
+   If no match is found in deny rules, or there are no deny rules execute next step.
+2. Check for a match in allow rules, if a match is found, the request will be allowed. 
+   Note that allow rules is a required field here, if not present, policy is invalid.
+3. If no match is found in allow rules, we deny the request.
+
+In the following policy example
+- Peer identity from ["spiffe://foo.com/sa/admin1", "spiffe://foo.com/sa/admin2"] is 
+  authorized to access any RPC methods in pkg.service
+- Any authenticated user is allowed to access "foo" and "bar" RPC methods.
+- Nobody can access "secret" RPC method, not even the admins as deny rules are evaluated 
+  first.
+
+```json
+{
+	"name": "example-policy",
+	"allow_rules": [
+		{
+			"name": "admin-access",
+			"source": {
+				"principals": [
+					"spiffe://foo.com/sa/admin1",
+					"spiffe://foo.com/sa/admin2"
+				]
+			},
+			"request": {
+				"paths": ["/pkg.service/*"]
+			}
+		},
+		{
+			"name": "dev-access",
+			"source": {
+				"principals": [
+					"*"
+				]
+			},
+			"request": {
+				"paths": [
+					"/pkg.service/foo",
+					"/pkg.service/bar"
+				]
+			}
+		}
+	],
+	"deny_rules": [
+		{
+			"name": "deny-access",
+			"request": {
+				"paths": [
+					"*/secret"
+				]
+			}
+		}
+	]
+}
+```
+
+As mentioned previously, internally gRPC Authorization engine consumes Envoy RBAC 
+proto. The user provided JSON SDK authorization policy will be translated to Envoy 
+RBAC proto and provided to Engines for evaluation. Note that SDK authorization policy 
+is a subset of Envoy RBAC, and it does not support all the fields that are present in 
+Envoy RBAC.
+
+### API
+
+gRPC will support both static initialization and dynamically reloading the policy
+from filesystem. In static initialization, the policy will be provided as a JSON
+string. In dynamic file reloading, the application will specify the file path that
+contains the authorization policy in JSON format.
```

Comment:
What happens when the Dynamic file loader reads invalid file contents? Do you keep the previous policy in force? Or do you retry the read in that case? Also how do you prevent contention with the writer of the policy?

Reply:
There are two cases to consider here -
1. If initial file reload fails - we won't create the policy provider, and simply return error to application
2. if file reload fails on refresh - then we skip that update, and use previously valid policy to make authorization decisions.

About the contention, if say the writer is updating the file at the same time that the thread reads from it, then it probably depends on when the file completed update. It is possible that when policy is updated, we won't immediately start using it. It may get taken up in the next refresh interval.

Reply:
> It is possible that when policy is updated, we won't immediately start using it. It may get taken up in the next refresh interval.

That's okay, I wasn't worried about missing the update. I was wondering if you are going to have extra code to guard against IO errors due to contention.

Reply:
For the MVP, we are not doing anything extra to guard against IO contention.